### PR TITLE
Refactoring out redundant enums

### PR DESCRIPTION
**Pull request by mattiecnvr**
Working on the scraper this was bothering me more and more. We have known for a while that the enum wrappers around traits were useless as the code stands now, so this rips them out so there is not another layer of indirection. This makes the codebase more idiomatic, but more importantly it makes it less redundant such that every change to the traits does not also require useless code to call the internal function with the exact same parameters.

I did toy with the idea of using static dispatch but I don't think the performance benefit is worth the headache at this time because of how we are constructing the values from configs. This is where I got stuck ultimately and while I think we could get everything except the inboxes to be static dispatch, it is probably not worth the hassle. It would certainly require more refactoring to how we init.

It is my hope that at one point the enum wrappers served a grater use than the do now, but at this point the literal only thing they did differently was for Inbox and Outbox they added support the `checkpoint` function with the mock implementation. This I solved by adding `commit` to the trait, documenting the use case, and giving a default implementation.

Changes:
- Redundant enums wrapping traits were removed.
- Removed `struct MyWrapper(Arc<T>)` wrappers and instead now favors just saying `Arc<T>`.
- Changed a few return types to be clone-able references rather than assuming a clone is wanted.
- Added `auto_impl` for most of the `abacus-base` traits to generate `Box` and `Arc` implementations.
- Allowed IDE to auto-reorg imports.
- Sorted implementation order of a few traits to match trait definitions.
- Pulled `checkpoint` function into the trait with a default `unimplemented` definition to support the mock implementations.

### CODE REVIEW

**Comment by mattiecnvr** [comment]
Force pushes didn't change anything but cleaned up commits and rebased on main.

**Review by tkporter — APPROVED** [review]
main question is around the `checkpoint` fns, but lgtm!

File: rust/abacus-base/src/settings/chains.rs [hunk 1]
```diff
@@ -105,28 +104,25 @@ impl ChainSetup<OutboxAddresses> {
         &self,
         signer: Option<Signers>,
         metrics: &CoreMetrics,
-    ) -> Result<Outboxes, Report> {
+    ) -> eyre::Result<Box<dyn Outbox>> {
```
[the remark this answers is not in the document]
Comment (mattiecnvr) [reply rust/abacus-base/src/settings/chains.rs:107]: Yeah, there's a normal Result elsewhere in the file iirc, and to be honest, with types that shadow std prelude ones I kinda prefer leaving one scope layer to differentiate.

Willing to change if you care, won't mind if it gets changed next time someone touches the file either.

File: rust/abacus-core/src/traits/inbox.rs [hunk 2]
```diff
@@ -20,4 +22,10 @@ pub trait Inbox: AbacusCommon + Send + Sync + Debug {
 
     /// The on-chain address of the inbox contract.
     fn contract_address(&self) -> Address;
+
+    /// Calls checkpoint on mock variant. Should only be used during tests.
+    #[auto_impl(keep_default_for(Box, Arc))]
+    fn checkpoint(&mut self) {
```
[the remark this answers is not in the document]
Comment (mattiecnvr) [reply rust/abacus-core/src/traits/inbox.rs:28]: I have no idea and wasn't actually able to find where it was being defined in the mocks, but it compiles so I probably just missed it.

I can try removing and see if anything cares.

**Pull request merged**